operator/v1/types_dns.go

```diff
@@ -162,6 +172,87 @@ type ForwardPlugin struct {
 	Policy ForwardingPolicy `json:"policy,omitempty"`
 }
 
+// UpstreamResolvers defines a schema for configuring the CoreDNS forward plugin in the
+// specific case of the default (".") server.
+// It defers from ForwardPlugin in the default values it accepts:
+// * At least one upstream should be specified.
+// * the default policy is Sequential
+type UpstreamResolvers struct {
+	// Upstreams is a list of resolvers to forward name queries for the "." domain.
+	// Each instance of CoreDNS performs health checking of Upstreams. When a healthy upstream
+	// returns an error during the exchange, another resolver is tried from Upstreams. The
+	// Upstreams are selected in the order specified in Policy.
+	//
+	// A maximum of 15 upstreams is allowed per ForwardPlugin.
+	// If no Upstreams are specified, /etc/resolv.conf is used by default
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=15
+	// +kubebuilder:default:={"type":"SystemResolvConf"}
+	Upstreams []Upstream `json:"upstreams"`
+
+	// Policy is used to determine the order in which upstream servers are selected for querying.
+	// Any one of the following values may be specified:
+	//
+	// * "Random" picks a random upstream server for each query.
+	// * "RoundRobin" picks upstream servers in a round-robin order, moving to the next server for each new query.
+	// * "Sequential" tries querying upstream servers in a sequential order until one responds, starting with the first server for each new query.
+	//
+	// The default value is "Sequential"
+	//
+	//+optional
+	// +kubebuilder:default:="Sequential"
+	Policy ForwardingPolicy `json:"policy,omitempty"`
+}
+
+// Upstream can either be of type SystemResolvConf, or of type Network.
+//
+// * For an Upstream of type SystemResolvConf, no further fields are necessary:
+//   The upstream will be configured to use /etc/resolv.conf.
+// * For an Upstream of type Network, a NetworkResolver field needs to be defined
+//   with an IP address or IP:port if the upstream listens on a port other than 53.
+//
+// +union
+type Upstream struct {
+
+	// Type defines whether this upstream contains an IP/IP:port resolver or the local /etc/resolv.conf.
+	// Type accepts 2 possible values: SystemResolvConf or Network.
+	//
+	// * When SystemResolvConf is used, the Upstream structure does not require any further fields to be defined:
+	//   /etc/resolv.conf will be used
+	// * When Network is used, the Upstream structure must contain at least an Address
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	Type UpstreamType `json:"type"`
```

Comment:
I believe the union feature is still not implemented for CRDs, only in server-side-apply. Hence, we have to specify that Address and Port are necessary if type is Network. Something along these lines:

```
properties:
  type: ....
  address: ...
  port: ...
anyOf:
- properties:
    type:
      enum: ["SystemResolvConf"]
   not:
     required: ["address", "port"]
- properties:
    type:
      enum: ["Network"]
  required: ["address", "port"]
```

Reply:
That's brilliant ! thanks @sttts !
I was looking for a way to define Address and port as conditionally required.
I didn't think of anyOf :+1: